packages/typescript/internal/ts_project.bzl

```diff
@@ -96,10 +97,17 @@ def _ts_project_impl(ctx):
     inputs = ctx.files.srcs + depset(transitive = deps_depsets).to_list() + [ctx.file.tsconfig]
     if ctx.attr.extends:
         inputs.extend(ctx.files.extends)
-    outputs = ctx.outputs.js_outs + ctx.outputs.map_outs + ctx.outputs.typings_outs + ctx.outputs.typing_maps_outs
+
+    json_outs = ctx.outputs.json_outs
+
+    # If there are no predeclared json_outs (probably bc of no .json or no outdir),
```

Comment:
this comment leaves an unanswered question: if there are no .json srcs and no .outdir then why do we need to declare any json_outs?
Is this load-bearing in some way? I pulled the change and your new test passes without this, so either remove it or make the test show why it's needed and update the comment.

Reply:
so the predeclared `json_outs` outputs are only computed when both `outdir` is there as an `attr` & `srcs` contains `json`.
- If there's no `json`, there's nothing to do.
- If there're some `json`s but `outdir` isn't there, `tsc` basically does a `copy_to_bin`-like behavior which is why the test passes for the `no-outdir` case (it grabs the src `json` instead of the copied one to compare?)

Reply:
I've updated the test that should demo this :)

Reply:
it still feels like this code is in the wrong place, compensating for json_outs having been passed into this function with the wrong value. Can it be fixed at the callsite instead? (line 402)
Seems like it just needs to match how js_outs is computed and passed in?

Reply:
initially I was doing just `_join(outdir, f for f in srcs if f.endswith(".json"))` but if `outdir` is `None` then I got `input & output are the same`. This is what special about `json` input :( For `.ts` u always get a `.js` output file but for `json` it's the same file if `outdir` isn't specified. Any rec on how to fix this?

Reply:
Right, makes sense, it's like copy_to_bin so you can't pre declare labels for outputs. What if we never declare them? (Degrade to common case for when outdir specified)?

Reply:
Yeah I tried just `declare_file` instead of splitting btwn predeclared outputs & `declare_file` but I can't get the path concat right for `outdir`, the file path in `ctx.files.srcs` seems to be the full path so I need some path mangling logic that I'm not sure if my assumptions are correct. Any guidance there?

Reply:
ah, I see, yes at this point the File object doesn't give you the path of the input relative to the package so you have to calculate it yourself
`src.dirname[len(ctx.label.package)+1:]`
I should have a commit to add here shortly...

Reply:
awesome thanks a lot!